### PR TITLE
Add flush command to ipfs files

### PR DESCRIPTION
**Pull request by whyrusleeping**

### CODE REVIEW

File: core/commands/files/files.go [hunk 1]
```diff
@@ -600,6 +614,40 @@ Examples:
...
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		nd, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		// take the lock and defer the unlock
+		defer nd.Blockstore.PinLock()()
```
Comment (jbenet) [inline core/commands/files/files.go]: this notation is a bit hard to read. i think these may be better:

```go
nd.Blockstore.PinLk.Lock()
defer nd.Blockstore.PinLk.Unlock()
// or
defer nd.Blockstore.PinLock().Unlock()
```

File: core/core.go [hunk 2]
```diff
@@ -474,6 +474,23 @@ func (n *IpfsNode) loadBootstrapPeers() ([]peer.PeerInfo, error) {
 func (n *IpfsNode) loadFilesRoot() error {
 	dsk := ds.NewKey("/local/filesroot")
 	pf := func(ctx context.Context, k key.Key) error {
+		ds := n.Repo.Datastore()
+		if old, err := ds.Get(dsk); err == nil && old != nil {
+			_ = n.Pinning.Unpin(ctx, key.Key(old.([]byte)), true)
```
Comment (jbenet) [inline core/core.go]: if this is an error and it is silenced then we may have a pin-leak (and pins growing and growing).

File: core/core.go [hunk 3]
```diff
@@ -474,6 +474,23 @@ func (n *IpfsNode) loadBootstrapPeers() ([]peer.PeerInfo, error) {
...
+		}
+		nnd, err := n.DAG.Get(ctx, k)
+		if err != nil {
+			return err
+		}
+
+		if err := n.Pinning.Pin(ctx, nnd, true); err != nil {
```
Comment (jbenet) [inline core/core.go]: it looks like no lock is acquired between the unpin + pin. this is unsafe and GC might happen in between.

Comment (whyrusleeping) [inline core/core.go]: good point.

**Comment by jbenet** [comment]
@whyrusleeping this looks good. is it _safe_ ?

**Comment by whyrusleeping** [comment]
@jbenet its too safe. I need to remove the pinning.

**Comment by jbenet** [comment]
> its too safe. I need to remove the pinning.

ahhaahah

**Comment by whyrusleeping** [comment]
alright, this is ready. most of the review is needed on the last commit, i ripped the mfs file apart into an 'inode' type object, and a 'file descriptor' type object. This makes working with file things much easier, and i've been meaning to do it for a long time. Things should be a little more performant, way more threadsafe, the files api stuff shouldnt hang for @diasdavid, and fuse should be much happier in general.

**Comment by Kubuxu** [comment]
From command info I can see that it causes changes not to be flushed to disk. Wouldn't it be better to call it `--no-flush`?

**Comment by whyrusleeping** [comment]
@Kubuxu the default is that changes are flushed to disk. the existing `--flush` command defaults (when not specified) to true. If you want to not flush changes to disk, you specify `--flush=false`

**Comment by Kubuxu** [comment]
Ahh, ok. Sorry for that.

**Comment by whyrusleeping** [comment]
@Kubuxu no worries, must be a documentation issue, what can i change to make that more clear?

**Comment by Kubuxu** [comment]
Just [this](https://github.com/ipfs/go-ipfs/pull/2196/files#diff-1feff21640e90e71a5ec9ccec2650fbdR32) wording is unfortunate. It suggests that just `--flush` is unsafe.

**Comment by whyrusleeping** [comment]
@Kubuxu there, that seems a little less confusing to me

**Comment by whyrusleeping** [comment]
@RichardLitt adding 'ipfs files flush' in this PR.

**Comment by RichardLitt** [comment]
Thanks @whyrusleeping. Noted in https://github.com/ipfs/http-api-spec/issues/57

**Pull request merged**